src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp

```diff
@@ -0,0 +1,458 @@
+#include "PostgreSQLReplicationHandler.h"
+
+#if USE_LIBPQXX
+#include <DataStreams/PostgreSQLBlockInputStream.h>
+#include <Databases/PostgreSQL/fetchPostgreSQLTableStructure.h>
+#include <Storages/PostgreSQL/StoragePostgreSQLReplica.h>
+
+#include <Common/setThreadName.h>
+#include <DataStreams/copyData.h>
+#include <Interpreters/InterpreterInsertQuery.h>
+#include <Poco/File.h>
+
+
+namespace DB
+{
+
+static const auto reschedule_ms = 500;
+static const auto max_thread_work_duration_ms = 60000;
+
+
+PostgreSQLReplicationHandler::PostgreSQLReplicationHandler(
+    const std::string & database_name_,
+    const std::string & conn_str,
+    const std::string & metadata_path_,
+    std::shared_ptr<Context> context_,
+    const size_t max_block_size_,
+    bool allow_minimal_ddl_,
+    bool is_postgresql_replica_database_engine_,
+    const String tables_list_)
+    : log(&Poco::Logger::get("PostgreSQLReplicaHandler"))
+    , context(context_)
+    , database_name(database_name_)
+    , connection_str(conn_str)
+    , metadata_path(metadata_path_)
+    , max_block_size(max_block_size_)
+    , allow_minimal_ddl(allow_minimal_ddl_)
+    , is_postgresql_replica_database_engine(is_postgresql_replica_database_engine_)
+    , tables_list(tables_list_)
+    , connection(std::make_shared<PostgreSQLConnection>(conn_str))
+{
+    replication_slot = fmt::format("{}_ch_replication_slot", database_name);
+    publication_name = fmt::format("{}_ch_publication", database_name);
+
+    startup_task = context->getSchedulePool().createTask("PostgreSQLReplicaStartup", [this]{ waitConnectionAndStart(); });
+    startup_task->deactivate();
+
+    consumer_task = context->getSchedulePool().createTask("PostgreSQLReplicaStartup", [this]{ consumerFunc(); });
+    consumer_task->deactivate();
+}
+
+
+void PostgreSQLReplicationHandler::addStorage(const std::string & table_name, StoragePostgreSQLReplica * storage)
+{
+    storages[table_name] = storage;
+}
+
+
+void PostgreSQLReplicationHandler::startup()
+{
+    startup_task->activateAndSchedule();
+}
+
+
+void PostgreSQLReplicationHandler::waitConnectionAndStart()
+{
+    try
+    {
+        /// Will throw pqxx::broken_connection if no connection at the moment
+        connection->conn();
+
+        startSynchronization();
+    }
+    catch (const pqxx::broken_connection & pqxx_error)
+    {
+        LOG_ERROR(log, "Unable to set up connection. Reconnection attempt will continue. Error message: {}",
+                pqxx_error.what());
+
+        startup_task->scheduleAfter(reschedule_ms);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+void PostgreSQLReplicationHandler::shutdown()
+{
+    stop_synchronization.store(true);
+    consumer_task->deactivate();
```

Comment:
```suggestion
    startup_task->deactivate();
    consumer_task->deactivate();
```
Seems like `startup_task` is not stopped on shutdown if all connections tries failed.

Reply:
Fixed.